academic_observatory_workflows/workflows/pubmed_telescope.py

```diff
@@ -0,0 +1,1623 @@
+
+                # Remove primary keys in the old upsert file set if it is present in a newer upsert file.
+                set_old.difference_update(same_records)
+
+                if same_records:
+                    logging.info(
+                        (
+                            f"Found {len(same_records)} newer records in file: {transform_file.split('/')[-1]} vs {key_file_path.split('/')[-1]}."
+                            f" {len(set_old)} records left in {key_file_path.split('/')[-1]}"
+                        )
+                    )
+
+                # Relace with list of primary keys with old keys removed.
+                upsert_files[key_file_path] = list(set_old)
+
+            # Add newest updatefile to list of updatefiles and it's records.
+            upsert_files[transform_file] = records
+
+        merged_upsert_files = []
+        first_file_index = upsert_files_to_merge[0].file_index
+        last_file_index = upsert_files_to_merge[-1].file_index
+        for file_count, (transform_file, records_to_write) in enumerate(upsert_files.items()):
+            logging.info(f"Reading in file to pull out records - {transform_file}")
+            with open(transform_file) as f_in:
+                incoming_data = [json.loads(line) for line in f_in]
+
+            # Turn list of records to a dictionary for faster lookup.
+            records = {
+                str(record[self.entity_list["upsert"].pmid_location]["PMID"]): record for record in incoming_data
+            }
+
+            merged_output_file = datafile.merged_transform_file_path(
+                self.entity_list["upsert"].type, first_file_index, last_file_index, part_num=file_count + 1
+            )
+
+            logging.info(f"Writing {len(records_to_write)} records to file - {merged_output_file}")
+
+            # Write to compressed *.jsonl.gz as this will be uploaded to GCS and imported into BQ.
+            with gzip.open(merged_output_file, "w") as f_out:
+                for record_key in records_to_write:
+                    f_out.write(str.encode(json.dumps(records[record_key], cls=PubMedCustomEncoder) + "\n"))
+
+            merged_upsert_files.append(merged_output_file)
+
+        # Send list of merged upsert record files to the Xcom.
+        ti: TaskInstance = kwargs["ti"]
+        ti.xcom_push(key="merged_upsert_files", value=merged_upsert_files)
+
+    def upload_merged_upsert_records(self, release: PubMedRelease, **kwargs):
+        """Upload the merged upsert records to GCS."""
+
+        ti: TaskInstance = kwargs["ti"]
+        merged_upsert_files = ti.xcom_pull(key="merged_upsert_files")
+
+        success = gcs_upload_files(
+            bucket_name=self.cloud_workspace.transform_bucket,
+            file_paths=merged_upsert_files,
+        )
+
+        set_task_state(success, kwargs["ti"].task_id, release=release)
+
+    def bq_load_upsert_table(self, release: PubMedRelease, **kwargs):
```

Comment:
It would be good to make the upsert, delete and snapshot names consistent with other workflows:
* pubmed_backup -> pubmed_snapshotYYYYMMDD.
* delete -> pubmed_delete (has table name in front and not sharded). The table name is in front as a namespace for the delete and upsert tables when there are multiple tables. You can make the tables expire, that is a good addition.
* upsert -> pubmed_upsert (has table name in front and not sharded).

You might also to make the schema names more consistent with other workflows or this naming, e.g. deletes.json delete.json and maybe upserts.json as pubmed.json.